src/main/java/br/com/meli/grupo1/desafio_spring/services/PurchaseService.java

```diff
@@ -69,6 +76,15 @@ private void verifyProducts(List<Long> productsIdList){
             throw new UnregisteredProductException("Produto solicitado nao cadastrado");
         }
     }
+    /* Verifica se exite a quantidade de produtos suficiente , lança exceçao com mensagem*/
+    private void verifyHasQuantityEnough(List<Product> productsInStock, Map<Long, Integer> quantidades) {
```

Comment:
Fiz essa mudanca aqui, soh pra reduzir a complexidade do codigo para O(n), sempre que possivel eh legal evitar fazer for dentro de for (da pra replicar no update se quiser). Se quiser trocar uma ideia depois, e fica a vontade pra mudar se quiser ;)

Reply:
boto fé, ficou  melhor. Eu acabei fazendo do jeito mais fácil pq o horário do expediente já tinha acabado. Tinha visto esta estrutura de Map q vcs tinha feito, pensei em usar mas não tinha entendido 100% como funcionava tb e deixei pra lá. Acho q daria pra refatorar tb pra otimizar o código pra reduzir a quantidade de vezes que os produtos são iterados, tipo fazer as verificações  de quantidade no momento que os produtos são selecionados de uma vez ou usar um Map para arquivar os produtos pra poder ter acesso direto pelo id na chave, mas ai teria que mexer em funções que vcs criaram e eu preferi não fazer isso agora.

Reply:
Verdade boa ideia, ficaria bem melhor com esse Map direto com os produtos!